pkg/controller/instanceset/instance_util.go

```diff
@@ -236,18 +241,67 @@ func GenerateAllInstanceNames(parentName string, replicas int32, templates []Ins
 	return instanceNameList
 }
 
-func GenerateInstanceNamesFromTemplate(parentName, templateName string, replicas int32, offlineInstances []string) []string {
-	instanceNames, _ := GenerateInstanceNames(parentName, templateName, replicas, 0, offlineInstances)
+func GenerateAllInstanceNamesWithTemplatesIndexRanges(parentName string, replicas int32, templates []InstanceTemplate, offlineInstances []string, templatesIndexRanges []workloads.InstanceTemplateIndexRanges) ([]string, error) {
+	totalReplicas := int32(0)
+	instanceNameList := make([]string, 0)
+	for _, template := range templates {
+		replicas := template.GetReplicas()
+		indexPoints := make([]int32, 0)
+		var err error
+		if len(templatesIndexRanges) > 0 {
+			indexPoints, err = GetIndexPointsByTemplateName(templatesIndexRanges, template.GetName())
+			if err != nil {
+				return nil, err
+			}
+		}
+		names := GenerateInstanceNamesFromTemplate(parentName, template.GetName(), replicas, offlineInstances, indexPoints)
+		instanceNameList = append(instanceNameList, names...)
+		totalReplicas += replicas
+	}
+	if totalReplicas < replicas {
+		indexPoints := make([]int32, 0)
+		var err error
+		if len(templatesIndexRanges) > 0 {
+			indexPoints, err = GetIndexPointsByTemplateName(templatesIndexRanges, "")
+			if err != nil {
+				return nil, err
+			}
+		}
+		names := GenerateInstanceNamesFromTemplate(parentName, "", replicas-totalReplicas, offlineInstances, indexPoints)
+		instanceNameList = append(instanceNameList, names...)
+	}
+	getNameNOrdinalFunc := func(i int) (string, int) {
+		return ParseParentNameAndOrdinal(instanceNameList[i])
+	}
+	baseSort(instanceNameList, getNameNOrdinalFunc, nil, true)
+	return instanceNameList, nil
+}
+
+func GenerateInstanceNamesFromTemplate(parentName, templateName string, replicas int32, offlineInstances []string, indexPoints []int32) []string {
+	instanceNames := GenerateInstanceNames(parentName, templateName, replicas, 0, offlineInstances, indexPoints)
 	return instanceNames
 }
 
 // GenerateInstanceNames generates instance names based on certain rules:
 // The naming convention for instances (pods) based on the Parent Name, InstanceTemplate Name, and ordinal.
 // The constructed instance name follows the pattern: $(parent.name)-$(template.name)-$(ordinal).
+// If the user specifies indexRanges, then only use indexRanges to generate instance names.
 func GenerateInstanceNames(parentName, templateName string,
```

Comment:
Add a new function, e.g. GenerateInstanceNamesWithOrdinalRange, instead of modify the orignal one.

Reply:
and the orignal GenerateInstanceNames can be implemented as a wrapper of GenerateInstanceNamesWithOrdinalRange